src/main/java/io/wispforest/owo/serialization/Codeck.java

```diff
@@ -0,0 +1,348 @@
+                NbtIo.write(value, stream);
+
+                Codeck.BYTE_ARRAY.encode(serializer, stream.toByteArray());
+            } catch (IOException e){
+                LOGGER.error("Unable to serialize the given NbtElement into the given format!");
+                throw new RuntimeException(e);
+            }
+        }
+
+        @Override
+        public <E> NbtElement decode(Deserializer<E> deserializer) {
+            if(deserializer instanceof SelfDescribedDeserializer<E> selfDescribedDeserializer){
+                var nbtSerializerzer = new NbtSerializer();
+
+                nbtSerializerzer.writeAny(selfDescribedDeserializer.readAny());
+
+                return nbtSerializerzer.result();
+            }
+
+            byte[] array = Codeck.BYTE_ARRAY.decode(deserializer);
+
+            try {
+                ByteArrayDataInput stream = ByteStreams.newDataInput(array);
+
+                return NbtIo.read(stream, NbtTagSizeTracker.ofUnlimitedBytes());
+            } catch (IOException e){
+                LOGGER.error("Unable to deserialize the given format into the desired NbtElement!");
+                throw new RuntimeException(e);
+            }
+        }
+    };
+
+    Codeck<NbtCompound> COMPOUND = Codeck.NBT_ELEMENT.then(element -> ((NbtCompound) element), compound -> compound);
+
+    //--
+
+    Codeck<Identifier> IDENTIFIER = Codeck.STRING.then(Identifier::new, Identifier::toString);
+
+    Codeck<ItemStack> ITEM_STACK = COMPOUND.then(ItemStack::fromNbt, stack -> stack.writeNbt(new NbtCompound()));
+
+    Codeck<UUID> UUID = Codeck.ifAttr(Codeck.STRING.then(java.util.UUID::fromString, java.util.UUID::toString), SerializationAttribute.HUMAN_READABLE)
+            .orElse(Codeck.INT_ARRAY.then(Uuids::toUuid, Uuids::toIntArray));
+
+    Codeck<Date> DATE = Codeck.ifAttr(Codeck.STRING.then(s -> Date.from(Instant.parse(s)), date -> date.toInstant().toString()), SerializationAttribute.HUMAN_READABLE)
+            .orElse(Codeck.LONG.then(Date::new, Date::getTime));
+
+    Codeck<PacketByteBuf> PACKET_BYTE_BUF = Codeck.BYTE_ARRAY
+            .then(bytes -> {
+                var byteBuf = PacketByteBufs.create();
+
+                byteBuf.writeBytes(bytes);
+
+                return byteBuf;
+            }, byteBuf -> {
+                var bytes = new byte[byteBuf.readerIndex()];
+
+                byteBuf.readBytes(bytes);
+
+                return bytes;
+            });
+
+    Codeck<BlockPos> BLOCK_POS = Codeck
+            .ifAttr(
+                    StructCodeckBuilder.of(
+                            StructField.of("x", StructCodeck.INT, BlockPos::getX),
+                            StructField.of("y", StructCodeck.INT, BlockPos::getX),
+                            StructField.of("z", StructCodeck.INT, BlockPos::getX),
+                            BlockPos::new
+                    ),
+                    SerializationAttribute.HUMAN_READABLE
```

Comment:
Won't this result in hard to edit NBT?

Reply:
Maybe, I have no clue how such would be represented, to say the least. It was mostly just a guess